distributed/shuffle/multi_file.py

```diff
@@ -0,0 +1,272 @@
+import asyncio
+import contextlib
+import os
+import pathlib
+import pickle
+import shutil
+import time
+from collections import defaultdict
+
+from dask.sizeof import sizeof
+from dask.utils import parse_bytes
+
+from distributed.utils import log_errors
+
+
+class MultiFile:
+    """Accept, buffer, and write many small objects to many files
+
+    This takes in lots of small objects, writes them to a local directory, and
+    then reads them back when all writes are complete.  It buffers these
+    objects in memory so that it can optimize disk access for larger writes.
+
+    **State**
+
+    -   shards: dict[str, list[bytes]]
+
+        This is our in-memory buffer of data waiting to be sent to other workers.
+
+    -   sizes: dict[str, int]
+
+        The size of each list of shards.  We find the largest and send data from that buffer
+
+    State
+    -----
+
+    memory_limit: str
+        A maximum amount of memory to use, like "1 GiB"
+    max_connections: int
+        The maximum number of connections to have out at once
+    max_message_size: str
+        The maximum size of a single message that we want to send
+
+    Parameters
+    ----------
+    directory: pathlib.Path
+        Where to write and read data.  Ideally points to fast disk.
+    dump: callable
+        Writes an object to a file, like pickle.dump
+    load: callable
+        Reads an object from that file, like pickle.load
+    join: callable
+        Joins many objects together
+    send: callable
+        How to send a list of shards to a worker
+    sizeof: callable
+        Measures the size of an object in memory
+    """
+
+    memory_limit = parse_bytes("1 GiB")
+    queue: asyncio.Queue = None
+    concurrent_files = 2
+
+    def __init__(
+        self,
+        directory,
+        dump=pickle.dump,
+        load=pickle.load,
+        join=None,
+        sizeof=sizeof,
+    ):
+        assert join
+        self.directory = pathlib.Path(directory)
+        if not os.path.exists(self.directory):
+            os.mkdir(self.directory)
+        self.dump = dump
+        self.load = load
+        self.join = join
+        self.sizeof = sizeof
+
+        self.shards = defaultdict(list)
+        self.sizes = defaultdict(int)
+        self.total_size = 0
+        self.total_received = 0
+
+        self.condition = asyncio.Condition()
+
+        self.bytes_written = 0
+        self.bytes_read = 0
+
+        self._done = False
+        self._futures = set()
+        self.active = set()
+        self.diagnostics = defaultdict(float)
+
+        if MultiFile.queue is None:
+            MultiFile.queue = asyncio.Queue()
```

Comment:
this asyncio.Queue is global and so you cannot use MultiFile instances from different threads

Reply:
Yup.  I'm not.  These are all created from within the event loop.  When the scheduler extension closes it sets this back to None in case other workers show up in the same process  but in a different thread (which would be a very stange occurence).

I think that I'm actually safe here for the moment, at least in non-pathological situations, but it's not the greatest solution.  I tacked on shared state just last night.  This part of the system isn't as thought through as the rest.

Reply:
This was cleaned up significantly